spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
> I don't think that "Both scans hook into exactly the same place in Spark code-base.
There is nothing to guess about it, you'd know if you checked Spark code base to see where and how those Scans are used:

https://github.com/apache/spark/blob/85842760dc4616577162f44cc0fa9db9bd23bd9c/sql/core/src/main/scala/org/apache/spark/sql/sources/DataSourceStrategy.scala#L35-57

`Expression`/`Filter` difference not matter during plan execution. They are used only during plan preparation stage, to construct an RDD, which is called once per job. The actual reading of data happens in `CassandraTableScanRDD`. If both `CatalystScan` and `PrunedFilteredScan` are producing exactly the same `CassandraTableScanRDD` (and they do), there is no performance difference and one of those scans is completely redundant. 

As for why `CatalystScan` is provided in some Spark internal APIs - the answer is easy - because they are internal Spark classes, and `CatalystScan` is an internal API, not Developer API. They are free to use it, as they control all the code. External packages like ours should not rely on internal Spark APIs for no reason. In this case you failed to convince me there is any valid reason, so please remove the class. Redundant code is increases cognitive load on code readers (e.g. reviewers), users (more options => more confusion which option to use) and reliance on internal APIs increases the risk of future compatibility problems.